domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,174 @@
+  *  MediaPlayer must be inactive when calling this function */
+  fun initializeMediaPlayer(url: String) {
+    if (mediaPlayerActive) throw MediaPlayerAlreadyInitException()
+    mediaPlayerActive = true
+    setMediaPlayerListeners()
+    prepareDataSource(url)
+  }
+
+  /* Call function to change data source of media player
+  *  Puts media player in a preparing state */
+  fun changeDataSource(url: String) {
+    prepared = false
+    stopUpdatingSeekBar()
+    mediaPlayer.reset()
+    prepareDataSource(url)
+  }
+
+  private fun setMediaPlayerListeners() {
+    mediaPlayer.setOnCompletionListener {
+      stopUpdatingSeekBar()
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.COMPLETED, 0, mediaPlayer.duration)))
+    }
+    mediaPlayer.setOnPreparedListener {
+      prepared = true
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, it.duration)))
+    }
+  }
+
+  private fun prepareDataSource(url: String) {
+    mediaPlayer.setDataSource(context, Uri.parse(url))
+    mediaPlayer.prepareAsync()
+    playState.postValue(AsyncResult.pending())
+  }
+
+  /** Call function to play audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in paused state */
+  fun play() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (!mediaPlayer.isPlaying) {
+      mediaPlayer.start()
+      scheduleNextSeekBarUpdate()
+    }
+  }
+
+  /** Call function to pause audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in playing state */
+  fun pause() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (mediaPlayer.isPlaying) {
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PAUSED, mediaPlayer.currentPosition, mediaPlayer.duration)))
+      mediaPlayer.pause()
+      stopUpdatingSeekBar()
+    }
+  }
+
+  private fun scheduleNextSeekBarUpdate() {
+    if (observerActive && prepared) {
+      seekBarLock.withLock {
+        nextUpdateJob = CoroutineScope(backgroundDispatcher).launch {
+          delay(SEEKBAR_UPDATE_FREQUENCY)
+          updateSeekBar()
+          scheduleNextSeekBarUpdate()
+        }
+      }
+    }
+  }
+
+  private fun updateSeekBar() {
+    if (mediaPlayer.isPlaying) {
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PLAYING, mediaPlayer.currentPosition, mediaPlayer.duration)))
+    }
+  }
+
+  private fun stopUpdatingSeekBar() {
+    seekBarLock.withLock {
+      nextUpdateJob?.cancel()
+      nextUpdateJob = null
+    }
+  }
+
+  /* Call function to release media player and stop seek bar updates
+  *  MediaPlayer must be active when calling this function */
+  fun releaseMediaPlayer() {
+    if (!mediaPlayerActive) throw MediaPlayerNotInitException()
+    mediaPlayerActive = false
+    prepared = false
+    mediaPlayer.release()
```

Comment:
Does this also remove the listeners? If not, maybe we should do that here, as well.

Reply:
Controller now takes in the fragment to remove observers based on it